pkg/network/proxy/proxy.go

```diff
@@ -161,35 +177,34 @@ func (proxy *OsdnProxy) handleAddOrUpdateNetNamespace(obj, _ interface{}, eventT
 	netns := obj.(*networkv1.NetNamespace)
 	klog.V(5).Infof("Watch %s event for NetNamespace %q", eventType, netns.Name)
 
-	proxy.idLock.Lock()
-	defer proxy.idLock.Unlock()
-	proxy.ids[netns.Name] = netns.NetID
+	proxy.Lock()
+	defer proxy.Unlock()
+
+	ns := proxy.getNamespace(netns.Name)
+	ns.global = (netns.NetID == network.GlobalVNID)
 }
 
 func (proxy *OsdnProxy) handleDeleteNetNamespace(obj interface{}) {
 	netns := obj.(*networkv1.NetNamespace)
 	klog.V(5).Infof("Watch %s event for NetNamespace %q", watch.Deleted, netns.Name)
 
-	proxy.idLock.Lock()
-	defer proxy.idLock.Unlock()
-	delete(proxy.ids, netns.Name)
-}
-
-func (proxy *OsdnProxy) isNamespaceGlobal(ns string) bool {
-	proxy.idLock.Lock()
-	defer proxy.idLock.Unlock()
+	proxy.Lock()
+	defer proxy.Unlock()
 
-	if proxy.ids[ns] == network.GlobalVNID {
-		return true
+	ns := proxy.namespaces[netns.Name]
+	if ns == nil {
+		return
 	}
-	return false
+
+	ns.global = false
```

Comment:
if we want to delete a global namespace do we have to force it to be no global?

Reply:
If we want `proxy.maybeDeleteNamespace` to garbage-collect the namespace, it has to be marked as non-global (since global namespaces are always interesting and so can't be GC'ed). But a namespace with no associated NetNamespace can't be global, so when a global namespace's NetNamespace is deleted, we mark it as being no-longer global, so that it can potentially be cleaned up.